### DISCUSSION TITLE
building dnn using meta-tegra opencv_4.8%.bbappend

### DISCUSSION THREAD

**Discussion by caveat-emptor** [OE4T/meta-tegra#1400] [Q&A]
Our devs have decided to use GRPC and they want the latest version. GRPC requires protobufs.

Our Yocto build is old and will build protobufs but the version is too old.

After some investigation I saw that that it is opencv that uses protobuf, we had old versions of opencv recipes so we got old protobufs.

I have taken the latest meta-openembedded and metag-tegra recipes for opencv.

They built but DNN wasn't being built and we need that.

There is a line in meta-tegra's opencv_4.5.5%bbappend
`OPENCV_CUDA_SUPPORT ?= "${@'cuda dnn' if opencv_cuda_flags(d) else ''}"`

The corresponding line in meta-tegra's opencv_4.8%bbappend is
`OPENCV_CUDA_SUPPORT ?= "${@'cuda' if opencv_cuda_flags(d) else ''}"`

The dnn isn't in 4.8.

So I hacked the line to add dnn and ran bitbake.

The build fails whilst building the libopencv_dnn.so library.

I see two possibilities:

1. I haven't done enough to get the build working
2. The dnn was removed from opencv_4.8%.bbappend because it no longer builds.

Does anyone know if there is an, inherent, problem with building dnn these days?  Known issues? Known solutions?

Are there other things that should be changed to get the build to work?

My log.do_compile is massive but I did see this near the bottom:

`FAILED: lib/libopencv_dnn.so.4.8.0 `

and this at the end:

```
ninja: build stopped: subcommand failed.
WARNING: /workdir/build/tmp/work/armv8a_tegra186-insphere-linux/opencv/4.8.0-r0/temp/run.do_compile.3199:182 exit 1 from 'eval ${DESTDIR:+DESTDIR=${DESTDIR} }VERBOSE=1 cmake --build '/workdir/build/tmp/work/armv8a_tegra186-insphere-linux/opencv/4.8.0-r0/build' "$@" -- ${EXTRA_OECMAKE_BUILD}'
WARNING: Backtrace (BB generated script): 
	#1: cmake_runcmake_build, /workdir/build/tmp/work/armv8a_tegra186-insphere-linux/opencv/4.8.0-r0/temp/run.do_compile.3199, line 182
	#2: cmake_do_compile, /workdir/build/tmp/work/armv8a_tegra186-insphere-linux/opencv/4.8.0-r0/temp/run.do_compile.3199, line 176
	#3: do_compile, /workdir/build/tmp/work/armv8a_tegra186-insphere-linux/opencv/4.8.0-r0/temp/run.do_compile.3199, line 171
	#4: main, /workdir/build/tmp/work/armv8a_tegra186-insphere-linux/opencv/4.8.0-r0/temp/run.do_compile.3199, line 195

```

**Comment by madisongh** [comment]
If you look at commit 35dab0f60478efe9d9ff44a072f8707c9257401f, where I disabled dnn, there are links there that point to the issues raised in opencv about the build errors that were getting reported.  Looks like the fix finally made it upstream, and the patch for it was [just added to meta-oe](https://github.com/openembedded/meta-openembedded/commit/225ce6a14a8110ab6b573b4dc9f5297a03d17e0f) last week.  So I've [reverted](https://github.com/OE4T/meta-tegra/commit/524bf0c79c2ae2549830367be373e9e01167084b) the change, and dnn is back again in the bbappend.   It does build cleanly with that patch in place (using master branch for OE-Core, meta-oe, and meta-tegra).

Actually digging into the compilation log to see what errors are being reported would be the best next step.  If you're trying to backport the newer opencv to an older branch, it could have to do with dependencies that might also need an update, or possibly the C++ compiler (since the build issues had to do with one of the dependencies for the opencv dnn module requiring C++17 support).
